internal/csi-addons/rbd/replication.go

```diff
@@ -548,6 +548,12 @@ func (rs *ReplicationServer) DemoteVolume(ctx context.Context,
 
 	// demote image to secondary
 	if info.IsPrimary() {
+		// If volume is demoted first, ControllerUnpublishVolume will fail to remove metadata
+		// since the image will be Read-only.
+		// Ensure volume is unpublished before demotion.
+		if err := checkVolumeUnpublished(ctx, rs.Driver.GetNodeID(), rbdVol); err != nil {
+			return nil, status.Errorf(codes.Internal, "volume cannot be demoted: %v", err)
```

Comment:
Returning `codes.Internal` seems incorrect, it should probably be `codes.FailedPrecondition` or `codes.Aborted`.

https://github.com/csi-addons/spec/blob/main/replication/README.md#error-scheme-3